segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,339 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
+                log.debug("{}: deleteGarbage - transaction is still active - re-queuing {}.", traceObjectId, infoToDelete.transactionId);
+                taskQueue.addTask(taskQueueName, infoToDelete);
+            } else {
+                val f = executeSerialized(() -> processTask(infoToDelete), infoToDelete.name);
+                if (null != f) {
+                    val now = currentTimeSupplier.get();
+                    if (infoToDelete.scheduledDeleteTime > currentTimeSupplier.get()) {
+                        futures.add(delaySupplier.apply(Duration.ofMillis(infoToDelete.scheduledDeleteTime - now))
+                                .thenComposeAsync(v -> f, storageExecutor));
+                    } else {
+                        futures.add(f);
+                    }
+                }
+            }
         }
+        return Futures.allOf(futures)
+                .thenRunAsync(() -> {
+                    queueSize.addAndGet(-1 * batch.size());
+                    SLTS_GC_TASK_PROCESSED.add(batch.size());
+                }, storageExecutor);
     }
 
     /**
-     * Delete the garbage chunks.
-     *
-     * This method retrieves a few eligible chunks for deletion at a time.
-     * The chunk is deleted only if the metadata for it does not exist or is marked inactive.
-     * If there are any errors then failed chunk is enqueued back up to a max number of attempts.
-     * If suspended or there are no items then it "sleeps" for time specified by configuration.
+     * Executes the given Callable asynchronously and returns a CompletableFuture that will be completed with the result.
+     * The operations are serialized on the segmentNames provided.
      *
-     * @param isBackground True if the caller is background task else False if called explicitly.
-     * @param maxItems     Maximum number of items to delete at a time.
-     * @return CompletableFuture which is completed when garbage is deleted.
-     */
-    CompletableFuture<Boolean> deleteGarbage(boolean isBackground, int maxItems) {
-        log.debug("{}: Iteration {} started.", traceObjectId, iterationId.get());
-        // Sleep if suspended.
-        if (suspended.get() && isBackground) {
-            log.info("{}: deleteGarbage - suspended - sleeping for {}.", traceObjectId, config.getGarbageCollectionDelay());
-            return CompletableFuture.completedFuture(false);
-        }
+     * @param operation    The Callable to execute.
+     * @param <R>       Return type of the operation.
+     * @param keyNames The names of the keys involved in this operation (for sequencing purposes).
+     * @return A CompletableFuture that, when completed, will contain the result of the operation.
+     * If the operation failed, it will contain the cause of the failure.
+     * */
+    private <R> CompletableFuture<R> executeSerialized(Callable<CompletableFuture<R>> operation, String... keyNames) {
```

Comment:
If you think that this could be a more general utility in other parts of the code, maybe you can think of moving it to `common` package. If not, I don't know if making these methods so generic is worth it, as I see that they are private and only called from one place.

Reply:
so far this very specific to GC/ChunkedSegmentStorage code.